src/utils.rs

```diff
@@ -23,6 +23,30 @@ impl ObjectIdentifier {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct BitString {
+    pub data: Vec<u8>,
+    pub bit_length: usize,
+}
+
+impl BitString {
+    pub fn new(data: Vec<u8>, bit_length: usize) -> Option<BitString> {
+        // TODO: there's got to be a good way to simplify this.
+        if data.is_empty() || bit_length == 0 {
+            if !(data.is_empty() && bit_length == 0) {
+                return None;
+            }
+        } else if (bit_length < (data.len() - 1) * 8) || (bit_length > data.len() * 8) {
+            return None;
+        }
```

Comment:
An alternative way of writing this:

```rust
match (data.len(), bit_length) {
    (0, 0) => (),
    (i, 0) if i > 0 => return None,
    (0, j) if j > 0 => return None,
    (i, j) if i * 8 < j => return None,
    (i, j) if (i - 1) * 8 > j => return None,
    _ => (),
};
```

I think it makes the possible control paths a little easier to understand, but up to you